torch/csrc/jit/codegen/cuda/runtime/memory.cu

```diff
@@ -0,0 +1,107 @@
+// Utility macro for this file
+#define DEVICE_INLINE __device__ inline
+
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 750))
+
+namespace Turing {
+
+namespace util {
+
+// Special utility for ldmatrix and cp_async
+DEVICE_INLINE unsigned toSmem(const void* ptr) {
+  unsigned smem_ptr;
+
+  asm("{ .reg .u64 smem_ptr; cvta.to.shared.u64 smem_ptr, %1; cvt.u32.u64 %0, smem_ptr; }\n"
+      : "=r"(smem_ptr)
+      : "l"(ptr));
+
+  return smem_ptr;
+}
+
+} // namespace util
+
+DEVICE_INLINE void ldMatrix(Array<__half, 4, 4>& out, void const* ptr) {
+  uint2& val = reinterpret_cast<uint2&>(out);
+  unsigned addr = util::toSmem(ptr);
+  asm volatile("ldmatrix.sync.aligned.x2.m8n8.shared.b16 {%0,%1}, [%2];"
+               : "=r"(val.x), "=r"(val.y)
+               : "r"(addr));
+}
+
+DEVICE_INLINE void ldMatrixT(Array<__half, 4, 4>& out, void const* ptr) {
+  uint2& val = reinterpret_cast<uint2&>(out);
+  unsigned addr = util::toSmem(ptr);
+  asm volatile("ldmatrix.sync.aligned.x2.trans.m8n8.shared.b16 {%0,%1}, [%2];"
+               : "=r"(val.x), "=r"(val.y)
+               : "r"(addr));
+}
+
+DEVICE_INLINE void ldMatrix(Array<__half, 8, 8>& out, void const* ptr) {
+  uint4& val = reinterpret_cast<uint4&>(out);
+  unsigned addr = util::toSmem(ptr);
+  asm volatile("ldmatrix.sync.aligned.x4.m8n8.shared.b16 {%0,%1,%2,%3}, [%4];"
+               : "=r"(val.x), "=r"(val.y), "=r"(val.z), "=r"(val.w)
+               : "r"(addr));
+}
+
+DEVICE_INLINE void ldMatrixT(Array<__half, 8, 8>& out, void const* ptr) {
+  uint4& val = reinterpret_cast<uint4&>(out);
+  unsigned addr = util::toSmem(ptr);
+  asm volatile(
+      "ldmatrix.sync.aligned.x4.trans.m8n8.shared.b16 {%0,%1,%2,%3}, [%4];"
+      : "=r"(val.x), "=r"(val.y), "=r"(val.z), "=r"(val.w)
+      : "r"(addr));
+}
+
+} // namespace Turing
+
+#endif // Arch 75
+
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 800))
+
+namespace Ampere {
+
+// MMA instruction wrappers (sm_80+):
+
+namespace util {
+
+// Special utility for cp_async
+DEVICE_INLINE unsigned toSmem(void* ptr) {
+  unsigned smem_ptr;
+
+  asm("{ .reg .u64 smem_ptr; cvta.to.shared.u64 smem_ptr, %1; cvt.u32.u64 %0, smem_ptr; }\n"
```

Comment:
Declare 64 bit register smem_ptr
Convert the input to a shared memory pointer
Convert to unsigned 32 bit pointer
Return the smem_ptr

Please rename smem_ptr either in the output, or in the ptr.